File: docs/python/RDKitCookbook/hybridization_type_and_count.py
from collections import defaultdict

from lillymol import *

def hybridization(unsaturation:int)->str:
  """Return a string for the Hybridziation of an atom with unsaturation
    Args:
      unsaturation: the difference btw nbonds and ncon for an atom
    Returns:
      String, SP, SP2, SP3
  """
  return ["SP3", "SP2", "SP"][unsaturation]

m = MolFromSmiles("CN1C=NC2=C1C(=O)N(C(=O)N2C)C#C")
total = defaultdict(int)

for i in range(0, len(m)):
  total[hybridization(m.unsaturation(i))] += 1

for (hyb,count) in total.items():
  print(f"{hyb} {count}")

# The atom object also has an unsaturation() method so we can access
# values via that means

total = defaultdict(int)
for atom in m:
  total[hybridization(atom.unsaturation())] += 1

for (hyb,count) in total.items():
  print(f"{hyb} {count}")

File: docs/JuLyMol/RDKitCookbook/hybridization_type_and_count.jl
using DataStructures

using LillyMol

# Return the Hybridization corresponding to `unsaturation`.
function hybridization(unsaturation)::String
  values = Array{String}([ "SP3", "SP2", "SP"])

  return values[unsaturation + 1]
end

function main()
  m = LillyMol.MolFromSmiles("CN1C=NC2=C1C(=O)N(C(=O)N2C)CC#C")

  # Accumulate the number of atoms with each hybridization type.
  total = DefaultDict{String, Integer}(0)
  for i in eachindex(m)
    total[hybridization(unsaturation(m, i))] += 1
  end

  for (h, count) in total
    println("$(h) $(count)")
  end

  # By getting the same information from the atoms.
  total = DefaultDict{String, Integer}(0)
  for atom in m
    total[hybridization(unsaturation(atom))] += 1
  end

  for (h, count) in total
    println("$(h) $(count)")
  end

  # By computing directly.
  total = DefaultDict{String, Int}(0)
  for atom in m
    total[hybridization(nbonds(atom) - ncon(atom))] += 1
  end

  for (h, count) in total
    println("$(h) $(count)")
  end
end

main()
